### PR TITLE
Refactor/Division

### LINKED ISSUE #15: Add Divisi

### LINKED ISSUE #16: Display Divisi

### LINKED ISSUE #17: Update Divisi

### PR DESCRIPTION
**Pull request by daffagrahito**
# Division

resolves #15, resolves #16, resolves #17

This pull request includes multiple changes to improve the handling of divisions in the repository and service layers, as well as their respective unit tests. The changes focus on enhancing the deletion, addition, and hierarchical retrieval of divisions.

### Improvements to division deletion:

* [`__tests__/unit/repository/division/delete-division.repository.unit.test.tsx`](diffhunk://#diff-67264ede083552929aef90f50e7940c58d9fe776c76908c0ce581a13ccd65640L22-R71): Added recursive calls to find and delete all descendants of a division and updated user `divisiId` to `null` for all affected users.
* [`__tests__/unit/repository/division/delete-division.repository.unit.test.tsx`](diffhunk://#diff-67264ede083552929aef90f50e7940c58d9fe776c76908c0ce581a13ccd65640L51-R97): Modified error handling to ensure proper user update and division deletion calls. [[1]](diffhunk://#diff-67264ede083552929aef90f50e7940c58d9fe776c76908c0ce581a13ccd65640L51-R97) [[2]](diffhunk://#diff-67264ede083552929aef90f50e7940c58d9fe776c76908c0ce581a13ccd65640L86-R140)

### Enhancements to division addition:

* [`__tests__/unit/services/division/add-division.service.unit.test.tsx`](diffhunk://#diff-e256ce0d9fd0c955ad26502d833471f8b8717d5251645f076aeddeb702d76b3eL82-R139): Added tests for handling `parentId` explicitly when adding a new division, including cases for `null`, `undefined`, and repository errors.
* [`src/controllers/division.controller.ts`](diffhunk://#diff-143bea0445d7e951c3517ca4ecdeb5c3adfda745bf2b2e86b4223b1c5ec1ee98L16-R24): Updated controller to handle `parentId` validation and set it to `null` if not provided.

### Improvements to division hierarchy retrieval:

* [`__tests__/unit/repository/division/get-division.repository.unit.test.tsx`](diffhunk://#diff-8da5525ea776c7112b9e06c97545c39b77bf026ba8d5920c454926af99c72aabL129-R337): Added tests to ensure proper hierarchical structure creation from a flat division list and handling of edge cases like missing parent IDs and circular references.
* [`src/repository/division.repository.ts`](diffhunk://#diff-9a83c88d3cd451d7ae775db651a8fec1c7022c9638b162588e547b9fb59374f8R31-R63): Refactored `getDivisionsHierarchy` method to use a bottom-up approach for building the hierarchy, ensuring efficient and accurate hierarchy creation.

## Summary by CodeRabbit

- **New Features**
  - Division management now removes all associated sub-divisions when a division is deleted and updates related user assignments.
  - The organizational hierarchy is built and presented more accurately, ensuring a clear view of divisions and their user counts.

- **Bug Fixes**
  - Enhanced input validation when adding divisions now delivers clearer error messages for missing or invalid parent identifiers.

### CODE REVIEW

**Comment by daffagrahito** [comment]
@coderabbitai review

File: src/repository/division.repository.ts [hunk 1]
```diff
@@ -159,26 +175,48 @@ class DivisionRepository {
...
+    }) || []; // Default to empty array if no children found
+  
+    const childIds = children.map((child) => child.id);
+  
+    const descendants = await Promise.all(
+      childIds.map((childId) => this.getAllChildrenIds(childId))
+    );
+  
+    return [...childIds, ...descendants.flat()];
+  }
```
[the remark this answers is not in the document]
Comment (daffagrahito) [reply src/repository/division.repository.ts:193]: will try later
